### PR TITLE
[Issue 930] Separate scale into scaleW and scaleB

### PR DESCRIPTION
**Pull request by plrectco**
## What changes were proposed in this pull request?
- add `scaleW` and `scaleB` to set the learning rate for weights and biases separately
- Create two new `Double` attributes `scaleW` and `scaleB` for `AbstractModule`. It does not work if the two attributes are defined as `T`
- Refactor `accGradParameters` such that it no longer need a `scale` argument
- Change the related layers accordingly

## How was this patch tested?

uni test

## Related links or issues (optional)
fixed https://github.com/intel-analytics/BigDL/issues/930  https://github.com/intel-analytics/BigDL/issues/336

### CODE REVIEW

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Euclidean.scala [hunk 1]
```diff
@@ -121,17 +121,16 @@ class Euclidean[T: ClassTag](val inputSize: Int, val outputSize: Int,
...
 
-  override def accGradParameters(input: Tensor[T], gradOutput: Tensor[T],
-    scale: Double = 1.0): Unit = {
+  override def accGradParameters(input: Tensor[T], gradOutput: Tensor[T]): Unit = {
 
     require(input.dim() == 1 || input.dim() == 2,
       "Euclidean: " + ErrorInfo.constrainInputAsVectorOrBatch)
     if (input.dim() == 1) {
-      gradWeight.add(ev.fromType(-scale), repeatBuffer)
+      gradWeight.add(ev.fromType(-ev.toType[Double](scaleW)), repeatBuffer)
```
Comment (qiuxin2012) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Euclidean.scala]: use ev.negative

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Euclidean.scala [hunk 2]
```diff
@@ -121,17 +121,16 @@ class Euclidean[T: ClassTag](val inputSize: Int, val outputSize: Int,
...
     } else if (input.dim() == 2) {
       sumBuffer.sum(repeatBuffer, 1)
       sumBuffer.resizeAs(weight)
-      gradWeight.add(ev.fromType(-scale), sumBuffer)
+      gradWeight.add(ev.fromType(-ev.toType[Double](scaleW)), sumBuffer)
```
Comment (qiuxin2012) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Euclidean.scala]: ev.negative

**Comment by qiuxin2012** [comment]
Please add more unit tests for the scaleW and scaleB.

**Comment by yiheng** [comment]
Can you add unit test for the changed layer?

**Comment by yiheng** [comment]
Please link the PR with the two related issues

**Comment by yiheng** [comment]
Give more introduction in the description of the PR

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Linear.scala [hunk 3]
```diff
@@ -136,16 +135,15 @@ class Linear[T: ClassTag](
...
       if (withBias) {
-        gradBias.add(value, gradOutput)
+        gradBias.add(ev.fromType[Double](scaleB), gradOutput)
       }
     }
     else if (input.dim() == 2) {
-      gradWeight.addmm(value, gradOutput.t, input)
-      if (withBias) gradBias.addmv(value, gradOutput.t, addBuffer)
+      gradWeight.addmm(ev.fromType[Double](scaleW), gradOutput.t, input)
+      if (withBias) gradBias.addmv(ev.fromType[Double](scaleB), gradOutput.t, addBuffer)
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Linear.scala]: This looks not right. We should scale the gradient after the regularizer process the gradient. @psyyz10 . Also please check other place where regularizer is used.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Bilinear.scala [hunk 4]
```diff
@@ -160,7 +160,7 @@ class Bilinear[T: ClassTag](
     gradInput
   }
 
-  override def accGradParameters(input: Table, gradOutput: Tensor[T], scale: Double = 1.0): Unit = {
+  override def accGradParameters(input: Table, gradOutput: Tensor[T]): Unit = {
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Bilinear.scala]: please check if scaleW and scaleB == 0, if so, just return or by pass the computing

File: spark/dl/src/test/scala/com/intel/analytics/bigdl/nn/SpatialConvolutionSpec.scala [hunk 5]
```diff
@@ -2669,4 +2669,154 @@ class SpatialConvolutionSpec extends FlatSpec with Matchers {
     checker.checkLayer[Double](layer, input) should be(true)
 
   }
+
+  "A SpatialConvolution with scaleW and scaleB" should "generate correct gradWeight gradBias" in {
```
Comment (qiuxin2012) [inline spark/dl/src/test/scala/com/intel/analytics/bigdl/nn/SpatialConvolutionSpec.scala]: make this UT simpler

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Utils.scala [hunk 6]
```diff
@@ -236,6 +236,22 @@ object Utils {
   }
 
   /**
+    * get whether the module is layerwise scaled
+    * @param model input module
+    * @return whether the module is layerwise scaled
+    */
+  def isLayerwiseScaled[T](model: Module[T]): Boolean = model match {
```
Comment (qiuxin2012) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Utils.scala]: works on Graph?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala [hunk 7]
```diff
@@ -711,6 +711,7 @@ class DistriOptimizer[T: ClassTag] (
     state("warmupIterationNum") = warmupIterationNum
     state("computeThresholdbatchSize") = computeThresholdbatchSize
     state("maxDropPercentage") = maxDropPercentage
+    state("isLayerwiseScaled") = Utils.isLayerwiseScaled(_model)
```
Comment (qiuxin2012) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala]: Also add this to LocalOptimizer

**Comment by qiuxin2012** [comment]
Please fix the conflict

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala [hunk 8]
```diff
@@ -62,6 +62,31 @@ abstract class AbstractModule[A <: Activity: ClassTag, B <: Activity: ClassTag,
...
   var gradInput: A = Activity[A, T]()
 
+  /*
+  * The scale of gradient weight and gradient bias
+  * before gradParameters being accumulated.
+  */
+  protected var scaleW: Double = 1.0
+  protected var scaleB: Double = 1.0
+
+  def getScaleW(): Double = {
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala]: Please add java docs for the four public methods.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Utils.scala (hunk 6 above)
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Utils.scala]: should work

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SpatialFullConvolution.scala [hunk 9]
```diff
@@ -510,46 +510,49 @@ class SpatialFullConvolution[A <: Activity : ClassTag, T: ClassTag](
...
-
-    // Do GEMV (note: this is a bit confusing because gemv assumes column-major matrices)
-    if (null != gradBias) {
-      ev.gemv(
-        'T',
-        k, m,
-        scale,
-        gradOutput.storage().array(), gradOutput.storageOffset() - 1, k,
-        ones.storage().array(), ones.storageOffset() - 1, 1,
+    if (scaleW != 0) {
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SpatialFullConvolution.scala]: scaleW is a T, does this really work?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SpatialFullConvolution.scala [hunk 10]
```diff
@@ -510,46 +510,49 @@ class SpatialFullConvolution[A <: Activity : ClassTag, T: ClassTag](
...
+        scaleW,
+        columns.storage().array(), columns.storageOffset() - 1, k,
+        input.storage().array(), input.storageOffset() - 1, k,
         ev.one,
-        gradBias.storage().array(), gradBias.storageOffset() - 1, 1
+        gradWeight.storage().array(), gradWeight.storageOffset() - 1, n
       )
     }
+
+    if (scaleB != 0) {
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SpatialFullConvolution.scala]: scaleB is a T, does this really work?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SpatialShareConvolution.scala [hunk 11]
```diff
@@ -220,7 +219,8 @@ class SpatialShareConvolution[T: ClassTag](
           gradWeightMM.select(1, g + 1),
           gradBias.narrow(1, g * nOutputPlane / nGroup + 1, nOutputPlane / nGroup),
           fInput.select(1, g + 1),
-          ev.fromType[Double](scale))
+          ev.fromType[Double](scaleW),
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SpatialShareConvolution.scala]: anywhere to check scale == 0?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Regularizer.scala [hunk 12]
```diff
@@ -140,13 +147,15 @@ class L1L2Regularizer[T: ClassTag](
...
    */
   private def accL2Regularization(
     alpha: Double,
     parameter: Tensor[T],
-    gradParameter: Tensor[T]
+    gradParameter: Tensor[T],
+    scale: Double
   ): Unit = {
-    if (alpha != 0) gradParameter.add(ev.fromType(alpha), parameter)
+    if (alpha != 0) gradParameter.add(ev.fromType(alpha* scale), parameter)
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Regularizer.scala]: && scale != 0?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Regularizer.scala [hunk 13]
```diff
@@ -121,13 +126,15 @@ class L1L2Regularizer[T: ClassTag](
...
    */
   private def accL1Regularization(
     alpha: Double,
     parameter: Tensor[T],
-    gradParameter: Tensor[T]
+    gradParameter: Tensor[T],
+    scale: Double
   ): Unit = {
-    if (alpha != 0) gradParameter.add(ev.fromType(alpha),
+    if (alpha != 0) gradParameter.add(ev.fromType(alpha*scale),
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Regularizer.scala]: && scale != 0

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SpatialDilatedConvolution.scala [hunk 14]
```diff
@@ -424,7 +423,7 @@ class SpatialDilatedConvolution[T: ClassTag](
       DenseTensorBLAS.gemm[T](
         't', 'n',
         n, m, k,
-        ev.fromType[Double](scale),
+        ev.fromType[Double](scaleW),
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SpatialDilatedConvolution.scala]: skip if scale == 0?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SpatialDilatedConvolution.scala [hunk 15]
```diff
@@ -441,7 +440,7 @@ class SpatialDilatedConvolution[T: ClassTag](
         ev.gemv(
           't',
           k, m,
-          ev.fromType[Double](scale),
+          ev.fromType[Double](scaleB),
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SpatialDilatedConvolution.scala]: skip if scale == 0?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/BatchNormalization.scala [hunk 16]
```diff
@@ -547,11 +558,11 @@ class BatchNormalization[@specialized(Float, Double) T: ClassTag](
         }
 
         if (null != gradWeight) {
-          gradWeight(_f - 1 + gradWeightOffset) += scale * dotp * invstd
+          gradWeight(_f - 1 + gradWeightOffset) += scaleW * dotp * invstd
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/BatchNormalization.scala]: check scaleW != 0?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/BatchNormalization.scala [hunk 17]
```diff
@@ -547,11 +558,11 @@ class BatchNormalization[@specialized(Float, Double) T: ClassTag](
...
         }
 
         if (null != gradBias) {
-          gradBias(_f - 1 + gradBiasOffset) += scale * sum
+          gradBias(_f - 1 + gradBiasOffset) += scaleB * sum
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/BatchNormalization.scala]: check scaleB != 0?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/BatchNormalization.scala [hunk 18]
```diff
@@ -635,11 +646,11 @@ class BatchNormalization[@specialized(Float, Double) T: ClassTag](
...
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/BatchNormalization.scala]: check scaleB != 0?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/BatchNormalization.scala [hunk 19]
```diff
@@ -635,11 +646,11 @@ class BatchNormalization[@specialized(Float, Double) T: ClassTag](
         }
 
         if (null != gradWeight) {
-          gradWeight(_f - 1 + gradWeightOffset) += scale * dotp * invstd
+          gradWeight(_f - 1 + gradWeightOffset) += scaleW * dotp * invstd
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/BatchNormalization.scala]: check scaleW != 0?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SpatialFullConvolution.scala (hunk 9 above)
Comment (qiuxin2012) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SpatialFullConvolution.scala]: scaleW and scaleB are double type.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SpatialShareConvolution.scala (hunk 11 above)
Comment (plrectco) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SpatialShareConvolution.scala]: This layer calls the `accGradParametersFrame` in `SpatialConvolution.scala`, in which the scale has been checked.

**Comment by yiheng** [comment]
Please also update the python regularizer API as the PR #1008 has been merged.

**Comment by plrectco** [comment]
I think we don't need to update the `regularizer` in Python, because in python the `regularizer` is passed to the layer as argument. Then it enters the scope of Scala. The layer calls `accRegularization` with `scale` as argument. We don't need to initialize `regularizer` class with `scale`, and only pass `scale` to it when it does the computation.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala [hunk 20]
```diff
@@ -63,6 +63,49 @@ abstract class AbstractModule[A <: Activity: ClassTag, B <: Activity: ClassTag,
...
+    scaleB
+  }
+
+  /**
+   * Set the scale of gradientWeight
+   *
+   * @param w the value of the scale of gradientWeight
+   * @return this
+   */
+  def setScaleW(w: Double): this.type = {
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala]: should we override this in container? So user can easily set a global scale @qiuxin2012

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala [hunk 21]
```diff
@@ -63,6 +63,49 @@ abstract class AbstractModule[A <: Activity: ClassTag, B <: Activity: ClassTag,
...
+    this
+  }
+
+  /**
+   * Set the scale of gradientBias
+   *
+   * @param b the value of the scale of gradientBias
+   * @return this
+   */
+  def setScaleB(b: Double): this.type = {
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala]: should we override this in container? So user can easily set a global scale @qiuxin2012

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala (hunk 20 above)
Comment (qiuxin2012) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala]: Yes, I think so.

**Comment by yiheng** [comment]
LGTM @qiuxin2012

**Comment by yiheng** [comment]
jenkins passed

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Utils.scala [hunk 22]
```diff
@@ -237,6 +237,22 @@ object Utils {
   }
 
   /**
+   * get whether the module is layerwise scaled
+   * @param model input module
+   * @return whether the module is layerwise scaled
+   */
+  def isLayerwiseScaled[T](model: Module[T]): Boolean = model match {
```
Comment (qiuxin2012) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Utils.scala]: add UT for this
